securedrop/journalist.py

```diff
@@ -333,6 +349,162 @@ def edit_account():
+    """Get all defined submission tags"""
+    submission_tags = db_session.query(SubmissionLabelType).all()
+    return submission_tags
+
+
+def get_source_tags(source_id):
+    source = Source.query.filter(Source.id == source_id).first()
+    source_tags = db_session.query(SourceTag, SourceLabelType) \
+                            .join(SourceLabelType) \
+                            .filter(SourceTag.source_id == source.id) \
+                            .group_by(SourceLabelType.label_text).all()
+    return source_tags
+
+
+def get_submission_tags(submission_id):
+    submission = Submission.query.filter(Submission.id == submission_id).first()
+    submission_tags = db_session.query(SubmissionTag, SubmissionLabelType) \
+                            .join(SubmissionLabelType) \
+                            .filter(SubmissionTag.submission_id == submission.id) \
+                            .group_by(SubmissionLabelType.label_text).all()
+    return submission_tags
+
+
+def delete_source_label(label_id):
+    matching_label = SourceLabelType.query.filter(SourceLabelType.id == label_id).all()
+    for label in matching_label:
+        db_session.delete(label)
+    db_session.commit()
+
+
+def delete_submission_label(label_id):
+    matching_label = SubmissionLabelType.query.filter(SubmissionLabelType.id == label_id).all()
+    for label in matching_label:
+        db_session.delete(label)
+    db_session.commit()
+
+
+@app.route('/admin/delete_source_label_type/<int:tag_id>', methods=('POST',))
+@admin_required
+def admin_delete_source_label_type(tag_id):
+    delete_source_label(tag_id)
+    return redirect(url_for('admin_index'))
+
+
+@app.route('/admin/create_source_label_type/', methods=('POST',))
+@admin_required
+def admin_create_source_label_type():
+    text_to_display = request.form['text_to_display']
+    all_labels = [x.label_text for x in get_all_defined_source_tags()]
+    if text_to_display not in all_labels:
+        db_session.add(SourceLabelType(label_text=text_to_display))
+        db_session.commit()
+    else:
+        flash('Tag already exists!', 'source')
+    return redirect(url_for('admin_index'))
+
+
+@app.route('/admin/create_submission_label_type/', methods=('POST',))
+@admin_required
+def admin_create_submission_label_type():
+    text_to_display = request.form['text_to_display']
+    all_labels = [x.label_text for x in get_all_defined_submission_tags()]
+    if text_to_display not in all_labels:
+        db_session.add(SubmissionLabelType(label_text=text_to_display))
+        db_session.commit()
+    else:
+        flash('Tag already exists!', 'submission')
+    return redirect(url_for('admin_index'))
+
+
+@app.route('/admin/delete_submission_label_type/<int:tag_id>', methods=('POST',))
+@admin_required
+def admin_delete_submission_label_type(tag_id):
+    delete_submission_label(tag_id)
+    return redirect(url_for('admin_index'))
+
+
+def create_source_tag(source_id, label_id):
```

Comment:
During in-person code review, we discussed how this is not DRY - this basically re-implements the same CRUD functionality for SourceTags that was already implemented for SubmissionTags, but duplicates code instead of reusing it.

This is fine for now - DRY is a good principle but not an absolute one, and it's often easier to maintain code that hasn't tried to get too fancy and abstracted in the service of avoiding repetition.

That said, this is a common problem and there are a number of elegant solutions to it. In Django-land, the most common solution is generic views. For inspiration, check out these Django docs:

* [Class-based views](https://docs.djangoproject.com/en/1.10/topics/class-based-views/)
* [Generic views](https://docs.djangoproject.com/en/1.10/topics/class-based-views/generic-display/)

There are probably a ton of projects that do this kind of thing for Flask via extensions. This is probably worth researching further, but for now this code is ok and suitable for merge.